### DISCUSSION TITLE
Database does not exist error

### DISCUSSION THREAD

**Discussion by ghost** [SeaQL/sea-orm#941] [Q&A] [answered]
I've [installed postgres on manjaro](https://wiki.archlinux.org/title/PostgreSQL), then I've created the user wiki with password wiki.
```
❯ sudo -iu postgres
[postgres@arch-desktop ~]$ createuser -P wiki
[postgres@arch-desktop ~]$ psql
postgres=# \du
Role name
wiki
```

And when I try to run [this program](https://codeberg.org/LifeSymbiont/media_wiki/src/commit/e1687774d6ebdb9afacb9d8cad5ab3f6a44b90b0) I get the error
```
thread 'main' panicked at 'Connection Error: error returned from database: database "wiki" does not exist', src/main.rs:53:9
```

[According to this](https://www.sea-ql.org/sea-orm-tutorial/ch01-01-project-setup.html#connecting-to-the-database-server)
```
// Change this according to your database implementation,
// or supply it as an environment variable.
// the database URL string follows the following format:
// "protocol://username:password@host:port/database"
const DATABASE_URL: &str = "postgres://wiki:wiki@localhost:5432";
```
it shouldn't be using `wiki` as the database name.

[In the tutorial](https://www.sea-ql.org/sea-orm-tutorial/ch01-01-project-setup.html#creating-a-database), why is it creating the database after it's already connected to it? Shouldn't it check if the database exists and create it only if it doesn't?

If I create the database manually with
```
[postgres@arch-desktop ~]$ createdb media_wiki_db
postgres=# ALTER DATABASE media_wiki_db OWNER TO wiki;
```

and use this instead
```
const DATABASE_URL: &str = "postgres://wiki:wiki@localhost:5432/media_wiki_db";
```

I get the error
```
thread 'main' panicked at 'Execution Error: error returned from database: cannot drop the currently open database', src/main.rs:53:9
```

thread 'main' panicked at 'Execution Error: error returned from database: permission denied to create database', src/main.rs:53:9

If I use the postgres database
```
const DATABASE_URL: &str = "postgres://wiki:wiki@localhost:5432/postgres";
```

I get
```
thread 'main' panicked at 'Execution Error: error returned from database: permission denied to create database', src/main.rs:53:9
```

Probably missing `CREATEDB` permission
```
❯ sudo -iu postgres
[postgres@arch-desktop ~]$ createdb test
postgres=# ALTER USER wiki CREATEDB;
postgres=# ALTER DATABASE test OWNER TO wiki;
postgres=# \l
\q
Ctrl+D
```

Changed to `const DATABASE_URL: &str = "postgres://wiki:wiki@localhost:5432/test";`
```
❯ cargo run
thread 'main' panicked at 'Connection Error: error returned from database: database "test/media_wiki_db" does not exist', src/main.rs:53:9
```

What am I doing wrong? How do I connect to the database? Actually what I would like to do is drop the database, recreate it and connect. At least for now.

**Comment by ghost** [comment, marked as answer]
I had to solve this by using two databases
```
const DATABASE_URL: &str = "postgres://admin:1234@localhost:5432";
const POSTGRES_DATABASE_URL: &str = "postgres://admin:1234@localhost:5432/postgres";
```
since I couldn't drop the database while using it.
So in the end I connect first to the postgres database and then refresh and connect to the other database.

[Solution context](https://codeberg.org/LifeSymbiont/media_wiki/src/commit/870d13dc48af66d348b4795088883f0ced60a8f1)

This solution wouldn't work if I were to change the database engine which lowers the usefulness of using the ORM.

An example that could be used independently of the database engine would be nice.

**Comment by billy1624** [comment]
Hey @JediMaster25, welcome! And thanks for asking!! For the tutorial, it's targeting audience without much prior knowledge and without its database created. Hence, the connection string doesn't include database name. However, normally you'd have your database created. So, your connection string is most likely to be suffixed with database name.
